src/gpio.rs

```diff
@@ -653,957 +352,833 @@ where
                 /// Opaque PUPDR register
-                pub pupdr: PUPDR,
+                pub pupdr: PUPDR<$port_id>,
                 $(
-                    #[doc = concat!("Pin ", stringify!($PXi))]
-                    pub $pxi: $PXi<$MODE>,
+                    /// Pin
+                    pub $pxi: $PXi $(<$MODE>)?,
                 )+
             }
 
-            impl GpioExt for $GPIOX {
+            impl super::GpioExt for $GPIOX {
                 type Parts = Parts;
 
                 fn split(self, ahb: &mut AHB) -> Parts {
                     <$GPIOX>::enable(ahb);
                     <$GPIOX>::reset(ahb);
 
                     Parts {
-                        afrh: AFRH(()),
-                        afrl: AFRL(()),
+                        afrh: Afr(()),
+                        afrl: Afr(()),
                         moder: MODER(()),
-                        ospeedr: OSPEEDR(()),
                         otyper: OTYPER(()),
+                        ospeedr: OSPEEDR(()),
                         pupdr: PUPDR(()),
                         $(
-                            $pxi: $PXi {
-                                gpio: $Gpiox,
-                                index: U::<$i>,
-                                _mode: PhantomData,
-                            },
+                            $pxi: $PXi::new(),
                         )+
                     }
                 }
             }
 
-            /// Opaque AFRH register
-            pub struct AFRH(());
+            pub type $PXn<MODE> = super::PEPin<$port_id, MODE>;
 
-            impl Afr for AFRH {
-                #[inline]
-                fn afx(&mut self, i: u8, x: u8) {
-                    const BITWIDTH: u8 = 4;
-                    unsafe { crate::modify_at!((*$GPIOX::ptr()).afrh, BITWIDTH, i - 8, x as u32) };
-                }
-            }
+            $(
+                pub type $PXi<MODE = Input> = super::Pin<$port_id, $i, MODE>;
 
-            /// Opaque AFRL register
-            pub struct AFRL(());
+                $(
+                    impl<MODE> IntoAf<$A> for $PXi<MODE> { }
+                )*
+            )+
 
-            impl Afr for AFRL {
-                #[inline]
-                fn afx(&mut self, i: u8, x: u8) {
-                    const BITWIDTH: u8 = 4;
-                    unsafe { crate::modify_at!((*$GPIOX::ptr()).afrl, BITWIDTH, i, x as u32) };
-                }
-            }
+        }
 
-            /// Opaque MODER register
-            pub struct MODER(());
+        pub use $gpiox::{ $($PXi,)+ };
+    }
+}
 
-            r_trait! {
-                ($GPIOX, $gpioy::moder::MODER15_A, 2);
-                impl Moder for MODER {
-                    fn input { INPUT }
-                    fn output { OUTPUT }
-                    fn alternate { ALTERNATE }
-                    fn analog { ANALOG }
-                }
-            }
+#[cfg(feature = "gpio-f302")]
+gpio!(GPIOA, gpioa, PA, 'A', PAn, [
```

Comment:
The macros here are different to the previous macros, which were auto generated by the codegen tools inside this repo. 

Maybe you can adjust that tool to get your desired output. But I'm fine if you don't like to bother. `codegen` needs a cleanup / overhaul anyway. 

But did you autogenerate this or did you adjust these invocations by hand, basing on the previous implementation?